itou/utils/tokens.py

```diff
@@ -12,4 +20,90 @@ def generate_random_token():
         C50364
         etc.
     """
-    return secrets.token_hex(3).upper()
+    # Little arithmetic trick to get the correct length in both cases:
+    # - if n even (n=2k), len(token_hex(1 + n // 2)) == 2+2k == n+2
+    # - if n odd (n=2k+1), len(token_hex(1 + n // 2)) == 2+2k == n+1
+    # the final `[:n]` ensures we get the correct length (n)
+    # while calculating as little extra characters as possible.
+    return secrets.token_hex(1 + n // 2)[:n].upper()
+
+
+class SiaeSignupTokenGenerator:
+    """
+    Strategy object used to generate and check tokens for the secure
+    siae signup mechanism.
+    Heavily inspired from django PasswordResetTokenGenerator :
+    https://github.com/django/django/blob/master/django/contrib/auth/tokens.py
+    """
+
+    key_salt = "itou.utils.tokens.SiaeSignupTokenGenerator"
+    secret = settings.SECRET_KEY
+
+    def make_token(self, siae):
+        """
+        Return a token that can be used once to do a signup
+        for the given siae and is valid only for a limited time.
+        """
+        return self._make_token_with_timestamp(siae, self._num_seconds(self._now()))
+
+    def check_token(self, siae, token):
+        """
+        Check that a siae signup token is correct for a given siae.
+        """
+        if not (siae and token):
+            return False
+        # Parse the token
+        try:
+            timestamp_b36, _ = token.split("-")
+        except ValueError:
+            return False
+
+        try:
+            timestamp = base36_to_int(timestamp_b36)
+        except ValueError:
+            return False
+
+        # Check that the timestamp/uid has not been tampered with
+        if not constant_time_compare(
+            self._make_token_with_timestamp(siae, timestamp), token
+        ):
+            return False
+
+        # Check the timestamp is within limit.
+        if (
+            self._num_seconds(self._now()) - timestamp
+        ) > SIAE_SIGNUP_MAGIC_LINK_TIMEOUT:
+            return False
+
+        return True
+
+    def _make_token_with_timestamp(self, siae, timestamp):
+        # timestamp is number of seconds since 2001-1-1. Converted to base 36,
+        # this gives us a 6 digit string until about 2069.
+        timestamp_b36 = int_to_base36(timestamp)
+        hash_string = salted_hmac(
+            self.key_salt, self._make_hash_value(siae, timestamp), secret=self.secret
```

Comment:
Là j'ai un problème parce que tu utilises `sha1` qui est l'argument par défaut de `salted_hmac`. Pourquoi tu n'as pas utilisé `sha256` ? c.f. https://github.com/django/django/blob/aee0bebc2faf9c6de8211b05d5f1281dc016084f/django/contrib/auth/tokens.py#L14

Reply:
Lol parce que ce changement est tout récent, du coup il n'y était pas à l'époque! :-D 

Ce commit a été mergé le 12 février :-)

![image](https://user-images.githubusercontent.com/10533583/75800648-967b0e00-5d82-11ea-92e0-5daa05a19463.png)

Lien: https://github.com/django/django/commit/da4923ea87124102aae4455e947ce24599c0365b

Reply:
Je vais appliquer le même changement du coup.

Reply:
TIL `sha1` is vulnerable and `sha256` should be used instead https://www.tbs-certificates.co.uk/FAQ/en/sha256.html

Reply:
Le pb c'est que `salted_hmac` n'a pas encore son argument `algorithm` dans la version `2.2.10` ni même dans la `3.0.3` (j'ai vérifié). Bref pour faire ce changement, il faut attendre une prochaine release de django, upgrader django et alors on pourra. @kemar 

![image](https://user-images.githubusercontent.com/10533583/75863763-b0abff00-5e09-11ea-9cfd-34176a2675cc.png)

![image](https://user-images.githubusercontent.com/10533583/75864129-3d56bd00-5e0a-11ea-871a-c33b7212f95b.png)

Reply:
Là pour le coup je suis désolé c'est de ma faute, j'ai vu `sha1`, j'ai flippé et je n'ai pas fait attention à la date du commit. Je te présente mes excuses.